auth/auth/flow.py

```diff
@@ -0,0 +1,114 @@
+import abc
+import msal
+import google.auth.transport.requests
+import google.oauth2.id_token
+import google_auth_oauthlib.flow
+import aiohttp.web
+import json
+import urllib.parse
+
+from gear.cloud_config import get_global_config
+
+
+class FlowResult:
+    def __init__(self, login_id: str, email: str, token: dict):
+        self.login_id = login_id
+        self.email = email
+        self.token = token
+
+
+class Flow(abc.ABC):
+    @abc.abstractmethod
+    def initiate_flow(self, redirect_uri: str) -> dict:
+        """
+        Initiates the OAuth2 flow. Usually run in response to a user clicking a login button.
+        The returned dict should be stored in a secure session so that the server can
+        identify to which OAuth2 flow a client is responding. In particular, the server must
+        pass this dict to :meth:`.receive_callback` in the OAuth2 callback.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def receive_callback(self, request: aiohttp.web.Request, flow_dict: dict) -> FlowResult:
+        """Concludes the OAuth2 flow by returning the user's identity and credentials."""
+        raise NotImplementedError
+
+
+class GoogleFlow(Flow):
+    scopes = [
+        'https://www.googleapis.com/auth/userinfo.profile',
+        'https://www.googleapis.com/auth/userinfo.email',
+        'openid',
+    ]
+
+    def __init__(self, credentials_file: str):
+        self._credentials_file = credentials_file
+
+    def initiate_flow(self, redirect_uri: str) -> dict:
+        flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+            self._credentials_file, scopes=self.scopes, state=None
+        )
+        flow.redirect_uri = redirect_uri
+        authorization_url, state = flow.authorization_url(access_type='offline', include_granted_scopes='true')
+
+        return {
+            'authorization_url': authorization_url,
+            'redirect_uri': redirect_uri,
+            'state': state,
+        }
+
+    def receive_callback(self, request: aiohttp.web.Request, flow_dict: dict) -> FlowResult:
+        flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+            self._credentials_file, scopes=self.scopes, state=flow_dict['state']
+        )
+        flow.redirect_uri = flow_dict['redirect_uri']
```

Comment:
This is a change from the previous behavior. This reuses the redirect uri from the initial initialize flow call. In the past, this would have been hard coded to `deploy_config.external_url('auth', '/oauth2callback')` in `/oauth2callback`

Reply:
Huh. Certainly this seems better than hardcoding it. It does seem like it needs to be set because its ultimately used in [`fetch_token`](https://github.com/requests/requests-oauthlib/blob/master/requests_oauthlib/oauth2_session.py#L321). 🤷‍♀️

Reply:
I bet it just has to be one of the valid redirect URIs for your application, but it doesn't matter which one. Or maybe requests-oauthlib is just unnecessarily sending extra data?